### ISSUE TITLE
Consider implementing new-array functions (ND namespace) as make-empty -> in-place assign

### ISSUE THREAD

**Issue by Ivorforce** [Ivorforce/NumDot#116]
We'll need to see if this affects binary size (it may) and runtime (it may not).

If it's a net plus, we can switch to it. If it's a tradeoff, it may be better than the current one where inplace assignments can be disabled.

**Comment by Ivorforce** [comment]
I think the benefits kinda depend on how specialized the function is. Array creation with `xexpression` parameter calls `semantic_base::assign` (with a hotfix for 0-D parameters), which just calls `assign_xexepression`, which then calls `rhs.assign_to(lhs)` if supported (i think that's just no-op views), otherwise `xexpression_assigner<tag>::assign_xexpression`. `assign_xexpression` of course contains the meat of the assignment:

```c++
        bool trivial_broadcast = resize(e1.derived_cast(), e2.derived_cast());
        base_type::assign_data(e1, e2, trivial_broadcast);
```

So, that's exactly as predicted: a resize, then data assign.

So we have two cases for our assignments:

- resize, then `assign_data`
- assert size, then `assign_data`

We gotta be careful about semantics, but I think with this background it may be possible to avoid some branching code and possibly reduce binary size while keeping runtime the same.

**Comment by Ivorforce** [comment]
Fixed in bcd5201e1bce98ae5ebb1b94eafae1029ce545f5.
